### PR TITLE
Adding fastp dockerfile to Staph-B

### PR DESCRIPTION
**Pull request by idolawoye**
This pull request is to include Fastp dockerfile to the Staph-B collection on docker-hub.

Pull Request (PR) checklist:
- [X] Include a description of what is in this pull request in this message.
- [X] The dockerfile successfully builds to a test target for the user creating the PR. (i.e. `docker build --tag samtools:1.15test --target test docker-builds/samtools/1.15` )
- [X] Directory structure as name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. `spades/3.12.0/Dockerfile`)
- [X] Main [README.md](https://github.com/StaPH-B/docker-builds/blob/master/README.md) has been updated to include the tool and/or version of the dockerfile(s) in this PR

Thank you!

### CODE REVIEW

**Comment by erinyoung** [comment]
First of all, you are my favorite person. This has been on my to-do list for a very long time.

File: fastp/0.23.2/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,33 @@
+FROM ubuntu:bionic as app
+
+ARG FASTP_VER="0.23.2"
+
+LABEL base.image="ubuntu:bionic"
+LABEL dockerfile.version="1"
+LABEL software="Fastp"
+LABEL software.version=$SOFTWARENAME_VER
```
Comment (erinyoung) [inline fastp/0.23.2/Dockerfile:8]: This line would probably be
```
LABEL software.version=$FASTP_VER
```

File: fastp/0.23.2/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+LABEL description="An ultra-fast all-in-one FASTQ preprocessor"
+LABEL website="https://github.com/OpenGene/fastp"
+LABEL license="https://github.com/OpenGene/fastp/blob/master/LICENSE"
+LABEL maintainer="Idowu Olawoye"
+LABEL maintainer.email="olawoyei0303@run.edu.ng"
+
+RUN apt-get update && apt-get install -y \
+ wget
```
Comment (erinyoung) [inline fastp/0.23.2/Dockerfile:16]: Could you also include `apt-get autoclean` and `rm -rf /var/lib/apt/lists/*`? This helps reduce the container size. Also, (while I'm remembering) could you add `ca-certificates`? This helps those that are using these containers with nextflow tower.
```
RUN apt-get update && apt-get install -y \
 wget \
 ca-certificates && \
 apt-get autoclean && rm -rf /var/lib/apt/lists/*
 ```

File: fastp/0.23.2/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN mkdir fastp && \
+ cd fastp && \ 
+ wget http://opengene.org/fastp/fastp.0.23.1 && \
+ mv fastp.0.23.1 fastp && \
+ chmod a+x ./fastp && \
+ mkdir /data
```
Comment (erinyoung) [inline fastp/0.23.2/Dockerfile:23]: Could you replace the hard-coded version with your variable? It helps when updating the file for future releases

Something like the following:

```
RUN mkdir fastp && \
 cd fastp && \ 
 wget http://opengene.org/fastp/fastp.${FASTP_VER} && \
 mv fastp.${FASTP_VER} fastp && \
 chmod a+x ./fastp && \
 mkdir /data
 ```

**Review by idolawoye — COMMENTED** [review]
Comments have now been resolved

**Review by erinyoung — CHANGES_REQUESTED** [review]
This looks great! But we do need you to make some changes before this can be merged.

To approve this request, you need to 
1 ) add yourself to `README.md` near the bottom as a contributor
2) adjust all the hardcoded versions in your dockerfile to your variable
3) add a test that uses fastp as it is intended to be used
4) add fastp to `Program_Licenses.md` (https://github.com/StaPH-B/docker-builds/blob/master/Program_Licenses.md)
5) add a small readme to fastp/0.23.2/README.md

The [template](https://github.com/StaPH-B/docker-builds/blob/master/.github/workflow-templates/readme-template.md) would suggests something like
```
# fastp container

Main tool : [fastp](https://github.com/OpenGene/fastp)


Full documentation: https://github.com/OpenGene/fastp

> A tool designed to provide fast all-in-one preprocessing for FastQ files.

# Example Usage

```
```
fastp -i sample_R1.fastq.gz -I sample_R2.fastq.gz -o sample_clean_PE1.fastq.gz -O sample_clean_PE2.fastq.gz 
```



Let us know if you need some assistance.

**Comment by erinyoung** [comment]
You're fast!

File: fastp/0.23.2/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+wget https://raw.githubusercontent.com/StaPH-B/docker-builds/master/tests/SARS-CoV-2/SRR13957123_2.fastq.gz
+
+# running fastp on the files
+fastp \
+        -i SRR13957123_1.fastq.gz \
+        -I SRR13957123_2.fastq.gz \
+        -o SRR13957123_PE1.fastq.gz \
+        -O SRR13957123_PE2.fastq.gz \
+        -h SRR13957123_fastp.html \
+        -j SRR13957123_fastp.json
```
[the remark this answers is not in the document]
Comment (idolawoye) [reply fastp/0.23.2/Dockerfile:48]: AH my bad! I was eating with one hand whilst resolving the comments with the other haha

**Comment by erinyoung** [comment]
I feel like this used to run, but is now getting stuck at 

```
#0 0.073 /fastp//fastp: 2: /fastp//fastp: Syntax error: newline unexpected
------
Dockerfile:35
--------------------
  33 |     FROM app as test
  34 |     
  35 | >>> RUN fastp --help
  36 |     
  37 |     # downloading sars-cov-2 fastq test files
--------------------
error: failed to solve: process "/bin/sh -c fastp --help" did not complete successfully: exit code: 2
Error: buildx failed with: error: failed to solve: process "/bin/sh -c fastp --help" did not complete successfully: exit code: 2
```

File: fastp/0.23.2/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ wget \
+ ca-certificates && \
+ apt-get autoclean && rm -rf /var/lib/apt/lists/*
+
+RUN mkdir fastp && \
+ cd fastp && \ 
+ wget http://opengene.org/fastp/fastp.{FASTP_VER} && \
+ mv fastp.{FASTP_VER} fastp && \
+ chmod a+x ./fastp && \
+ mkdir /data
```
Comment (erinyoung) [inline fastp/0.23.2/Dockerfile:25]: I think I found the issue! {FASTP_VER} _should_ be ${FASTP_VER}

**Comment by erinyoung** [comment]
This is related to https://github.com/StaPH-B/docker-builds/pull/403 and https://github.com/StaPH-B/docker-builds/pull/402

**Comment by erinyoung** [comment]
And the container is up on dockerhub at https://hub.docker.com/r/staphb/fastp/tags

Thank you for your contributions!!!

**Pull request merged**